refactor(core): explore approach to avoid storing LView in __ngContext__

crisbeto opened the pull request:
**Note:** this is mostly an exploratory PR to check whether the approach actually works and to start a discussion. There's more work to be done to make it mergeable (e.g. it introduces new circular dependencies, there are no unit tests etc.).

Explores adding a unique ID to each `LView` which is stored in `__ngContext__`, rather than the LView itself. This will avoid worsening memory leaks where a detached element is retained along with its `LView`.

Furthermore, it reworks `LContext` to store the ID of its `LView`, rather than the `LView` itself, because the `LContext` can also be stored in `__ngContext__`.

crisbeto commented on goldens/circular-deps/packages.json:
```diff
@@ -108,6 +108,33 @@
     "packages/compiler/src/render3/view/styling_builder.ts",
     "packages/compiler/src/render3/view/template.ts"
   ],
+  [
```
Please ignore these circular dependencies. If we decide to go for this approach, I'll make sure that no new circular dependencies are introduced.

crisbeto commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -123,6 +123,8 @@ function renderChildComponents(hostLView: LView, components: number[]): void {
   }
 }
 
+const ACTIVE_LVIEWS: (LView|null)[] = [];
```
This part is potentially leaky, but IMO it'll be fine, because this is the only place where `LView`s are created and destroyed and we have full control over it.

mhevery reviewed (commented):
This Looks promising, I think we should explore further. My thoughts:
- With this we are essentially getting into the business of memory management. In practice this is a ref counting scheme. The concern with ref-counting schemes is that one may forget to release memory.
- I think the key observation here is that when `LView` gets destroyed we already have to keep track of it because we need to call `ngDestroy` life-cycle events. This means that once destroyed there is no way to resurrect the data. Piggybacking to clean up `ACTIVE_LVIEWS` would be a minor cost. 
- There is still a concern that someone may detach `ViewRef` and than just drop it without proper `ngDestroy`. This would create a memory leak, but a good argument can be made that it is app error since the `ViewRef` could be re-added and dropping it creates an unreachable references. 

NOTE:
- Did not review the code in depth only looked at the main algorithm for tracking ids
- Will need a way to deal with holes in the `ACTIVE_LVIEWS` and also to ensure that we are not creating holey arrays.

crisbeto committed:
refactor(core): explore approach to avoid storing LView in __ngContext__

**Note:** this is mostly an exploratory PR to check whether the approach actually works and to start a discussion. There's more work to be done to make it mergeable (e.g. it introduces new circular dependencies, there are no unit tests etc.).

Explores adding a unique ID to each `LView` which is stored in `__ngContext__`, rather than the LView itself. This will avoid worsening memory leaks where a detached element is retained along with its `LView`.

Also reworks `LContext` to store the ID of its `LView`, rather than the `LView` itself, because the `LContext` can also be stored in `__ngContext__`.

crisbeto commented:
Closing in favor of #41358.
